service/history/workflow/workflow_test/mutable_state_impl_test.go

```diff
@@ -0,0 +1,363 @@
+	t.Parallel()
+
+	for _, tc := range []mutationTestCase{
+		{
+			name:              "numEvents<maxNumEvents",
+			numEvents:         1,
+			maxNumEvents:      2,
+			transactionPolicy: workflow.TransactionPolicyActive,
+			expectFailure:     false,
+		},
+		{
+			name:              "numEvents=maxNumEvents",
+			numEvents:         2,
+			maxNumEvents:      2,
+			transactionPolicy: workflow.TransactionPolicyActive,
+			expectFailure:     true,
+		},
+		{
+			name:              "numEvents>maxNumEvents",
+			numEvents:         3,
+			maxNumEvents:      2,
+			transactionPolicy: workflow.TransactionPolicyActive,
+			expectFailure:     true,
+		},
+	} {
+		t.Run(tc.name, tc.Run)
+	}
+}
+
+type mutationTestCase struct {
+	name              string
+	transactionPolicy workflow.TransactionPolicy
+	numEvents         int
+	maxNumEvents      int
+	expectFailure     bool
+}
+
+func (c mutationTestCase) Run(t *testing.T) {
+	t.Parallel()
+
+	nsEntry := tests.LocalNamespaceEntry
+	ms := c.createMutableState(t, nsEntry)
+
+	_, err := ms.AddWorkflowExecutionStartedEvent(
+		commonpb.WorkflowExecution{
+			WorkflowId: "694B31C3-1FDC-4C9E-87BC-747D539BF0CD",
+			RunId:      "3E1836B8-8692-440D-9495-45D9EABDED6B",
+		},
+		&historyservice.StartWorkflowExecutionRequest{
+			Attempt:     1,
+			NamespaceId: nsEntry.ID().String(),
+			StartRequest: &workflowservice.StartWorkflowExecutionRequest{
+				WorkflowType:        &commonpb.WorkflowType{Name: "workflow-type"},
+				TaskQueue:           &taskqueuepb.TaskQueue{Name: "task-queue-name"},
+				WorkflowRunTimeout:  timestamp.DurationPtr(200 * time.Second),
+				WorkflowTaskTimeout: timestamp.DurationPtr(1 * time.Second),
+			},
+		},
+	)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	c.completeOneWFT(t, ms)
+
+	wft, err := ms.AddWorkflowTaskScheduledEvent(false, enums.WORKFLOW_TASK_TYPE_NORMAL)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	_, wft, err = ms.AddWorkflowTaskStartedEvent(wft.ScheduledEventID, wft.RequestID, wft.TaskQueue, "")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	for i := 0; i < c.numEvents; i++ {
+		c.addWorkflowExecutionSignaled(t, i, ms)
+	}
+
+	_, workflowEvents, err := ms.CloseTransactionAsMutation(c.transactionPolicy)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if c.expectFailure {
+		c.testFailure(t, ms, wft, workflowEvents)
+	} else {
+		c.testSuccess(t, ms, workflowEvents)
+	}
+}
+
+func (c mutationTestCase) addWorkflowExecutionSignaled(t *testing.T, i int, ms *workflow.MutableStateImpl) {
+	t.Helper()
+
+	payload := &commonpb.Payloads{}
+	identity := fmt.Sprintf("%d", i)
+	header := &commonpb.Header{}
+
+	_, err := ms.AddWorkflowExecutionSignaled(
+		"signal-name",
+		payload,
+		identity,
+		header,
+		false,
+	)
+	if err != nil {
+		t.Fatal(err)
+	}
+}
+
+func (c mutationTestCase) createMutableState(t *testing.T, nsEntry *namespace.Namespace) *workflow.MutableStateImpl {
+	t.Helper()
+
+	ctrl := gomock.NewController(t)
+	shardContext := shard.NewMockContext(ctrl)
```

Comment:
there's a test implementation for shardContext you can use, and avoid some EXPECT() call like generating taskID or getting timeSource/metricsHandler/etc. .
check history/shard/context_testutil.go

Reply:
Done